### PR TITLE
Pagination

### PR DESCRIPTION
**Pull request by Jonasdoubleyou**

### CODE REVIEW

File: frontend/.eslintrc.js [hunk 1]
```diff
@@ -64,6 +64,7 @@ module.exports = {
     '@typescript-eslint/no-empty-interface': 'off',
     '@typescript-eslint/no-explicit-any': 'off',
     '@typescript-eslint/no-use-before-define': 'off',
+    '@typescript-eslint/explicit-function-return-type': 'off',
```
Comment (mgjm) [inline frontend/.eslintrc.js]: Can you move the rule to line `64`. The settings are sorted alphabetically.

File: frontend/src/contexts/Paging.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import * as React from 'react';
```
Comment (mgjm) [inline frontend/src/contexts/Paging.tsx]:
```suggestion
import React, { createContext, useContext, useState } from 'react';
```

File: frontend/src/contexts/Paging.tsx [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+/* Pagination */
+type IPage = '/' | '/bp';
+
+const PageContext = React.createContext<{ page: IPage; setPage(p: IPage): void }>({ page: '/', setPage() { } });
```
Comment (mgjm) [inline frontend/src/contexts/Paging.tsx]:
```suggestion
const PageContext = createContext<{ page: IPage; setPage(p: IPage): void }>({ page: '/', setPage() { } });
```

File: frontend/src/contexts/Paging.tsx [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+export const WithPagination = ({ children }: React.PropsWithChildren<{}>) => {
+    const [page, _setPage] = React.useState<IPage>(window.location.pathname as IPage);
```
Comment (mgjm) [inline frontend/src/contexts/Paging.tsx]:
```suggestion
    const [page, _setPage] = useState(window.location.pathname as IPage);
```

Comment (mgjm) [reply frontend/src/contexts/Paging.tsx]: Is there a reason why `_setPage` is prefixed with `_`?

File: frontend/src/contexts/Paging.tsx [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    function setPage(newPage: IPage) {
+        console.log('page', newPage);
+        window.history.pushState({ page }, 'Grenzexpress', newPage);
+        _setPage(newPage);
+    }
+    return <PageContext.Provider value={{ page, setPage }}>{children}</PageContext.Provider>;
+};
+
+export const Page = ({ children, name }: React.PropsWithChildren<{ name: IPage }>) => {
+    const { page } = React.useContext(PageContext);
```
Comment (mgjm) [inline frontend/src/contexts/Paging.tsx]:
```suggestion
    const { page } = useContext(PageContext);
```

File: frontend/src/contexts/Paging.tsx [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    if (page !== name) return null;
+    return <>{children}</>;
+};
+
+export const Link = ({ to, children }: React.PropsWithChildren<{ to: IPage }>) => {
+    const { setPage } = React.useContext(PageContext);
```
Comment (mgjm) [inline frontend/src/contexts/Paging.tsx]:
```suggestion
    const { setPage } = useContext(PageContext);
```

**Review by mgjm — CHANGES_REQUESTED** [review]

**Comment by mgjm** [comment]
Hint: You can run `npm run lint` to run the lint checks locally.

File: frontend/src/contexts/Paging.tsx (hunk 4 above)
Comment (Jonasdoubleyou) [reply frontend/src/contexts/Paging.tsx]: Cause `setPage` gets reimplemented later on. That's my personal coding style though ... I'm fine with another name

Comment (mgjm) [reply frontend/src/contexts/Paging.tsx]: Ok... then maybe leave a comment to make this clear.

Comment (Jonasdoubleyou) [reply frontend/src/contexts/Paging.tsx]: I guess as we only have two days we should not focus on code quality *that* much

**Comment by mgjm** [comment]
A general question: You implemented the pagination using "Higher-Order Components". I would propose to use "React-Hooks" instead.

Example:
```ts

export function usePage(): [IPage, (oage: IPage) => void] {
  const [page, _setPage] = React.useState<IPage>(window.location.pathname as IPage);

  React.useEffect(() => {
    window.addEventListener('popstate', (e) => {
      _setPage(e.state.page || '');
    });
  }, []);

  const setPage = (newPage: IPage) => {
    console.log('page', newPage);
    window.history.pushState({ page }, 'Grenzexpress', newPage);
    _setPage(newPage);
  };

  return [page, setPage];
}
```

**Comment by Jonasdoubleyou** [comment]
@mgjm but those effects are only ever used inside WithPagination. Don't see a sense in making that reusable

**Pull request merged**